apollo-router-core/src/plugin.rs

```diff
@@ -64,41 +64,59 @@ pub fn plugins() -> HashMap<String, PluginFactory> {
     PLUGIN_REGISTRY.lock().expect("Lock poisoned").clone()
 }
 
+/// All router plugins must implement the Plugin trait. This trait defines lifecycle hooks that enable hooking into Apollo Router services.
+/// The trait also provides a default implementations for each hook, which returns the associated service unmodified.
 #[async_trait]
 pub trait Plugin: Send + Sync + 'static + Sized {
     type Config: JsonSchema + DeserializeOwned;
 
+    /// This is invoked once after the router starts and compiled-in
+    /// plugins are registered.
     fn new(config: Self::Config) -> Result<Self, BoxError>;
 
-    // Plugins will receive a notification that they should start up and shut down.
+    /// Plugins will receive a notification that they should start up and shut down.
+    /// This is invoked whenever configuration is changed
+    /// during router execution, including at startup.
     async fn startup(&mut self) -> Result<(), BoxError> {
         Ok(())
     }
+
+    /// This is invoked whenever configuration is changed
+    /// during router execution, including at shutdown.
     async fn shutdown(&mut self) -> Result<(), BoxError> {
         Ok(())
     }
 
+    /// This service runs at the very beginning and very end of the request lifecycle.
+    /// Define router_service if your customization needs to interact at the earliest or latest point possible.
+    /// For example, this is a good opportunity to perform JWT verification before allowing a request to proceed further.
     fn router_service(
         &mut self,
         service: BoxService<RouterRequest, RouterResponse, BoxError>,
     ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
         service
     }
 
+    /// This service handles generating the query plan for each incoming request.
+    /// Define `query_planning_service` if your customization needs to interact with query planning functionality (for example, to log query plan details).
     fn query_planning_service(
         &mut self,
         service: BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError>,
     ) -> BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError> {
         service
     }
 
+    /// This service handles initiating the execution of a query plan after it's been generated.
+    /// Define `execution_service` if your customization includes logic to govern execution (for example, if you want to block a particular query based on a policy decision).
     fn execution_service(
         &mut self,
         service: BoxService<ExecutionRequest, ExecutionResponse, BoxError>,
     ) -> BoxService<ExecutionRequest, ExecutionResponse, BoxError> {
         service
     }
 
+    /// This service handles communication between the Apollo Router and your subgraphs.
+    /// Define `subgraph_service` to configure this communication (for example, to dynamically add headers to pass to a subgraph).
```

Comment:
Perhaps note that "_name" is the name of the subgraph service and can be used to customize processing by subgraph if desired? (same for DynPlugin)